tiny_dnn/core/kernels/conv2d_op_internal.h

```diff
@@ -72,41 +80,51 @@ inline void conv2d_op_internal(const tensor_t &in_data,
 }
 
 /******************************************************************/
+template <typename S1,
+          typename S2,
+          typename S3,
+          typename S4,
+          typename S5,
+          typename S6>
+inline void conv2d_op_internal(const Tensor<float_t, S1> &prev_out,
+                               const Tensor<float_t, S2> &weigths,
+                               Tensor<float_t, S3> &weights_grads,
+                               Tensor<float_t, S4> &bias_grads,
+                               Tensor<float_t, S5> &curr_delta,
+                               Tensor<float_t, S6> &prev_delta,
+                               const core::conv_params &params,
+                               const bool parallelize) {
+  for_i(parallelize, prev_out.shape()[0], [&](int sample) {
+    const float_t *prev = prev_out.host_pointer(sample, 0);
+    float_t *curr       = curr_delta.host_pointer(sample, 0);
+    float_t *prev_d     = prev_delta.host_pointer(sample, 0);
 
-template <typename tensor_t, typename vec_t>
-void conv2d_op_internal(const tensor_t &prev_out,
-                        const vec_t &W,
-                        tensor_t &dW,
-                        tensor_t &db,
-                        tensor_t &curr_delta,
-                        tensor_t &prev_delta,
-                        const core::conv_params &params,
-                        const bool parallelize) {
-  typedef typename vec_t::value_type float_t;
-
-  for_i(parallelize, prev_out.size(), [&](int sample) {
+    const float_t *weight_begin = weigths.host_pointer(0, 0);
     // propagate delta to previous layer
     for (serial_size_t inc = 0; inc < params.in.depth_; inc++) {
```

Comment:
As for now, Tensor's 2D: first dimension is samples and second is all the data. We can't obtain depth from it, we need 4D Tensor for that